### PR TITLE
replace int with operator.index for lax

### PR DESCRIPTION
**Pull request by YouJiacheng**
Note caller of _eye has ensured int.
Change with `jax.numpy.tri` is included in this PR for potential reverting

### CODE REVIEW

File: jax/_src/lax/lax.py [hunk 1]
```diff
@@ -1138,7 +1139,7 @@ def _delta(dtype: DType, shape: Shape, axes: Sequence[int]) -> Array:
 
 def _tri(dtype: DType, shape: Shape, offset: int) -> Array:
   """Like numpy.tri, create a 2D array with ones below a diagonal."""
-  offset = int(offset)
+  assert type(offset) is int
```
Comment (jakevdp) [inline jax/_src/lax/lax.py:1142]: Here and elsewhere: we shouldn't assert that inputs are of type int, because (for example), this fails for numpy integer scalars.

File: jax/_src/lax/lax.py [hunk 2]
```diff
@@ -1340,7 +1341,7 @@ def _iter(tracer):
   if tracer.ndim == 0:
     raise TypeError("iteration over a 0-d array")  # same as numpy error
   else:
-    n = int(tracer.shape[0])
+    n = operator.index(tracer.shape[0])
```
Comment (jakevdp) [inline jax/_src/lax/lax.py:1344]: Why is this change necessary? Is there any situation in which an element of a shape is not an integer?

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -2369,6 +2369,11 @@ def tri(N, M=None, k=0, dtype=None):
   lax_internal._check_user_dtype_supported(dtype, "tri")
   M = M if M is not None else N
   dtype = dtype or float32
+  try:
+    k = operator.index(k)
+  except TypeError:
+    warnings.warn("using k supports int(k) but not k.__index__, e.g. float will be deprecated.", DeprecationWarning)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:2375]: I don't think the meaning or intent of this warning will be clear to users who see it. How about:
```
warnings.warn(f"jnp.tri: k must be an integer. In future JAX versions non-integer values will result in an error.")
```

File: jax/_src/lax/utils.py [hunk 4]
```diff
@@ -115,3 +116,12 @@ def standard_named_shape_rule(*avals, **kwargs):
 
 def _standard_weak_type_rule(*avals, **kwargs):
   return all(aval.weak_type for aval in avals)
+
+def _int_replaced_by_index(x):
+  try:
+    return operator.index(x)
+  except TypeError:
+    msg = ("using types support int(x) but don't support x.__index__ as int, "
+          "e.g. float, np.float64 will be deprecated")
```
Comment (jakevdp) [inline jax/_src/lax/utils.py:125]: We should offer the user some clue here about the provenance of the error: imagine running a long & involved script and seeing several copies of this error, reported as coming from line 126 of `jax/_src/lax/utils.py`. What would you do to fix it?

Perhaps `_int_replaced_by_index` could accept the function name & variable name as strings, and put these in the warning to give context.

Alternatively, this may be an argument to inline this check in each place we use it, so that the warning is reported in the relevant context.

File: jax/_src/lax/control_flow.py [hunk 5]
```diff
@@ -205,8 +206,8 @@ def fori_loop(lower, upper, body_fun, init_val):
   if (isinstance(core.get_aval(lower), ConcreteArray) and
       isinstance(core.get_aval(upper), ConcreteArray)):
     try:
-      lower_ = int(lower)
-      upper_ = int(upper)
+      lower_ = _int_replaced_by_index(lower)
+      upper_ = _int_replaced_by_index(upper)
```
Comment (jakevdp) [inline jax/_src/lax/control_flow.py:210]: The intent here is to check whether `lower` and `upper` are tracers. We should not raise deprecation warnings here, and we probably should not change this to `operator.index` in the long run.

File: tests/lax_control_flow_test.py [hunk 6]
```diff
@@ -1969,7 +1969,7 @@ def loop(loop_impl, x):
     jtu.check_grads(loop_lax, (x,), order=2, modes=["fwd"])
 
   def testStaticForiGrad(self):
-    func = lambda x: lax.fori_loop(x, x + 2., lambda i, c: c, x)
+    func = lambda x: lax.fori_loop(int(x), int(x) + 2, lambda i, c: c, x)
```
Comment (jakevdp) [inline tests/lax_control_flow_test.py:1972]: I don't think this test change is necessary once you address the comments within the `fori_loop` implementation.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/lax/lax.py (hunk 1 above)
Comment (YouJiacheng) [reply jax/_src/lax/lax.py:1142]: I add a to `int` conversion in the only caller of `_tri` -- `jax.numpy.tri`, to be consistency with  `jax.numpy.eye`.

File: jax/_src/numpy/lax_numpy.py (hunk 3 above)
Comment (YouJiacheng) [reply jax/_src/numpy/lax_numpy.py:2375]: Thanks! I am not familiar with how to write a clear warning.

File: jax/_src/lax/control_flow.py (hunk 5 above)
Comment (YouJiacheng) [reply jax/_src/lax/control_flow.py:210]: Good catch! This is my mistake!

Comment (YouJiacheng) [reply jax/_src/lax/control_flow.py:210]: > The side effect here is that non-traced floating point array inputs would be switched from using the scan implementation to using the while_loop.implementation, which is not what we want.

What about using concrete check here?

File: jax/_src/lax/lax.py (hunk 2 above)
Comment (YouJiacheng) [reply jax/_src/lax/lax.py:1344]: Got it.

File: jax/_src/lax/control_flow.py [hunk 7]
```diff
@@ -205,9 +206,9 @@ def fori_loop(lower, upper, body_fun, init_val):
   if (isinstance(core.get_aval(lower), ConcreteArray) and
       isinstance(core.get_aval(upper), ConcreteArray)):
     try:
-      lower_ = int(lower)
-      upper_ = int(upper)
-    except TypeError:
+      lower_ = _check_integer("lax.fori_loop", "lower", lower)
+      upper_ = _check_integer("lax.fori_loop", "upper", upper)
+    except TypeError: # change to TracerIntegerConversionError?
```
Comment (YouJiacheng) [inline jax/_src/lax/control_flow.py:211]: @jakevdp  Could you please give some advice?
Does `fori_loop` not allow `Tracer` with `ConcreteArray` type `aval`?

Comment (jakevdp) [reply jax/_src/lax/control_flow.py:211]: I believe that `int(x)` will return the concrete value if `x` is a tracer with a concrete aval.

Comment (jakevdp) [reply jax/_src/lax/control_flow.py:211]: Again, as far as I can tell this code works correctly and does exactly what we want it to do, so unless you have found a bug somewhere I'd say it's not broken, and there's no need to change it.

Comment (YouJiacheng) [reply jax/_src/lax/control_flow.py:211]: > Again, as far as I can tell this code works correctly and does exactly what we want it to do, so unless you have found a bug somewhere I'd say it's not broken, and there's no need to change it.

Sorry to be dense about this point, I have little experience in working for a real project. -- BTW, I am curious about how to get a tracer with a concrete aval.

Comment (jakevdp) [reply jax/_src/lax/control_flow.py]: It will mainly come in via autodiff. For example:
```python
from jax import grad

def f(x):
  print(x.aval)
  return x

_ = grad(f)(1.0)
# ConcreteArray(1.0, dtype=float32, weak_type=True)
```

Comment (YouJiacheng) [reply jax/_src/lax/control_flow.py]: Thanks for nice example! -- And I find Tracer define appropriate `__int__` method.

File: jax/_src/lax/lax.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/lax/lax.py:1142]: There's little or no downside to repeating `offset = operator.index(offset)` in place of an `isinstance` assertion; the upside is that if this low-level function is ever used elsewhere, it will be better behaved.

**Comment by jakevdp** [comment]
Closing since this has bit-rotted – feel free to re-open if you wish to continue working on this!
